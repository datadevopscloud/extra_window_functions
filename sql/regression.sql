CREATE EXTENSION extra_window_functions;

CREATE TABLE things (
    part integer NOT NULL,
    ord integer NOT NULL,
    val integer
);

COPY things FROM stdin;
1	1	64664
1	2	8779
1	3	14005
1	4	57699
1	5	98842
1	6	88563
1	7	70453
1	8	82824
1	9	62453
2	1	\N
2	2	51714
2	3	17096
2	4	41605
2	5	15366
2	6	87359
2	7	98990
2	8	34982
2	9	3343
3	1	21903
3	2	24605
3	3	6242
3	4	24947
3	5	79535
3	6	66903
3	7	42269
3	8	31143
3	9	\N
4	1	\N
4	2	49723
4	3	23958
4	4	80796
4	5	\N
4	6	41066
4	7	72991
4	8	33734
4	9	\N
5	1	\N
5	2	\N
5	3	\N
5	4	\N
5	5	\N
5	6	\N
5	7	\N
5	8	\N
5	9	\N
\.

/* FLIP_FLOP */

SELECT part, ord, val,
       flip_flop(val % 2 = 0) OVER w AS flip_flop_1,
       flip_flop(val % 2 = 0, val % 2 = 1) OVER w AS flip_flop_2
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;

/* LAG */

SELECT part, ord, val,
       lag(val) OVER w AS lag,
       lag_ignore_nulls(val) OVER w AS lag_in,
       lag_ignore_nulls(val, 2) OVER w AS lag_in_off,
       lag_ignore_nulls(val, 2, -9999999) OVER w AS lag_in_off_d
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;

/* LEAD */

SELECT part, ord, val,
       lead(val) OVER w AS lead,
       lead_ignore_nulls(val) OVER w AS lead_in,
       lead_ignore_nulls(val, 2) OVER w AS lead_in_off,
       lead_ignore_nulls(val, 2, 9999999) OVER w AS lead_in_off_d
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;

/* FIRST_VALUE */

SELECT part, ord, val,
       first_value(val) OVER w AS fv,
       first_value_ignore_nulls(val) OVER w AS fv_in,
       first_value_ignore_nulls(val, 9999999) OVER w AS fv_in_d
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;

/* LAST_VALUE */

SELECT part, ord, val,
       last_value(val) OVER w AS lv,
       last_value_ignore_nulls(val) OVER w AS lv_in,
       last_value_ignore_nulls(val, -9999999) OVER w AS lv_in_d
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;

/* NTH_VALUE */

SELECT part, ord, val,
       nth_value(val, 3) OVER w AS nth,
       nth_value_ignore_nulls(val, 3) OVER w AS nth_in
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;

SELECT part, ord, val,
       nth_value(val, 3) OVER w AS nth,
       nth_value_from_last(val, 3) OVER w AS nth_fl
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;

SELECT part, ord, val,
       nth_value_from_last(val, 3) OVER w AS nth_fl,
       nth_value_from_last_ignore_nulls(val, 3) OVER w AS nth_fl_in
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;

DROP TABLE things;

DROP EXTENSION extra_window_functions;
