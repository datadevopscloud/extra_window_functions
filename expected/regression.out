CREATE EXTENSION extra_window_functions;
CREATE TABLE things (
    part integer NOT NULL,
    ord integer NOT NULL,
    val integer
);
COPY things FROM stdin;
/* FLIP_FLOP */
SELECT part, ord, val,
       flip_flop(val % 2 = 0) OVER w AS flip_flop_1,
       flip_flop(val % 2 = 0, val % 2 = 1) OVER w AS flip_flop_2
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;
 part | ord |  val  | flip_flop_1 | flip_flop_2 
------+-----+-------+-------------+-------------
    1 |   1 | 64664 | t           | t
    1 |   2 |  8779 | t           | t
    1 |   3 | 14005 | t           | f
    1 |   4 | 57699 | t           | f
    1 |   5 | 98842 | t           | t
    1 |   6 | 88563 | f           | t
    1 |   7 | 70453 | f           | f
    1 |   8 | 82824 | t           | t
    1 |   9 | 62453 | t           | t
    2 |   1 |       | f           | f
    2 |   2 | 51714 | t           | t
    2 |   3 | 17096 | t           | t
    2 |   4 | 41605 | f           | t
    2 |   5 | 15366 | t           | t
    2 |   6 | 87359 | t           | t
    2 |   7 | 98990 | t           | t
    2 |   8 | 34982 | t           | t
    2 |   9 |  3343 | t           | t
    3 |   1 | 21903 | f           | f
    3 |   2 | 24605 | f           | f
    3 |   3 |  6242 | t           | t
    3 |   4 | 24947 | t           | t
    3 |   5 | 79535 | t           | f
    3 |   6 | 66903 | t           | f
    3 |   7 | 42269 | t           | f
    3 |   8 | 31143 | t           | f
    3 |   9 |       | t           | f
    4 |   1 |       | f           | f
    4 |   2 | 49723 | f           | f
    4 |   3 | 23958 | t           | t
    4 |   4 | 80796 | t           | t
    4 |   5 |       | f           | t
    4 |   6 | 41066 | t           | t
    4 |   7 | 72991 | t           | t
    4 |   8 | 33734 | t           | t
    4 |   9 |       | f           | t
    5 |   1 |       | f           | f
    5 |   2 |       | f           | f
    5 |   3 |       | f           | f
    5 |   4 |       | f           | f
    5 |   5 |       | f           | f
    5 |   6 |       | f           | f
    5 |   7 |       | f           | f
    5 |   8 |       | f           | f
    5 |   9 |       | f           | f
(45 rows)

/* LAG */
SELECT part, ord, val,
       lag(val) OVER w AS lag,
       lag_ignore_nulls(val) OVER w AS lag_in,
       lag_ignore_nulls(val, 2) OVER w AS lag_in_off,
       lag_ignore_nulls(val, 2, -9999999) OVER w AS lag_in_off_d
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;
 part | ord |  val  |  lag  | lag_in | lag_in_off | lag_in_off_d 
------+-----+-------+-------+--------+------------+--------------
    1 |   1 | 64664 |       |        |            |     -9999999
    1 |   2 |  8779 | 64664 |  64664 |            |     -9999999
    1 |   3 | 14005 |  8779 |   8779 |      64664 |        64664
    1 |   4 | 57699 | 14005 |  14005 |       8779 |         8779
    1 |   5 | 98842 | 57699 |  57699 |      14005 |        14005
    1 |   6 | 88563 | 98842 |  98842 |      57699 |        57699
    1 |   7 | 70453 | 88563 |  88563 |      98842 |        98842
    1 |   8 | 82824 | 70453 |  70453 |      88563 |        88563
    1 |   9 | 62453 | 82824 |  82824 |      70453 |        70453
    2 |   1 |       |       |        |            |     -9999999
    2 |   2 | 51714 |       |        |            |     -9999999
    2 |   3 | 17096 | 51714 |  51714 |            |     -9999999
    2 |   4 | 41605 | 17096 |  17096 |      51714 |        51714
    2 |   5 | 15366 | 41605 |  41605 |      17096 |        17096
    2 |   6 | 87359 | 15366 |  15366 |      41605 |        41605
    2 |   7 | 98990 | 87359 |  87359 |      15366 |        15366
    2 |   8 | 34982 | 98990 |  98990 |      87359 |        87359
    2 |   9 |  3343 | 34982 |  34982 |      98990 |        98990
    3 |   1 | 21903 |       |        |            |     -9999999
    3 |   2 | 24605 | 21903 |  21903 |            |     -9999999
    3 |   3 |  6242 | 24605 |  24605 |      21903 |        21903
    3 |   4 | 24947 |  6242 |   6242 |      24605 |        24605
    3 |   5 | 79535 | 24947 |  24947 |       6242 |         6242
    3 |   6 | 66903 | 79535 |  79535 |      24947 |        24947
    3 |   7 | 42269 | 66903 |  66903 |      79535 |        79535
    3 |   8 | 31143 | 42269 |  42269 |      66903 |        66903
    3 |   9 |       | 31143 |  31143 |      42269 |        42269
    4 |   1 |       |       |        |            |     -9999999
    4 |   2 | 49723 |       |        |            |     -9999999
    4 |   3 | 23958 | 49723 |  49723 |            |     -9999999
    4 |   4 | 80796 | 23958 |  23958 |      49723 |        49723
    4 |   5 |       | 80796 |  80796 |      23958 |        23958
    4 |   6 | 41066 |       |  80796 |      23958 |        23958
    4 |   7 | 72991 | 41066 |  41066 |      80796 |        80796
    4 |   8 | 33734 | 72991 |  72991 |      41066 |        41066
    4 |   9 |       | 33734 |  33734 |      72991 |        72991
    5 |   1 |       |       |        |            |     -9999999
    5 |   2 |       |       |        |            |     -9999999
    5 |   3 |       |       |        |            |     -9999999
    5 |   4 |       |       |        |            |     -9999999
    5 |   5 |       |       |        |            |     -9999999
    5 |   6 |       |       |        |            |     -9999999
    5 |   7 |       |       |        |            |     -9999999
    5 |   8 |       |       |        |            |     -9999999
    5 |   9 |       |       |        |            |     -9999999
(45 rows)

/* LEAD */
SELECT part, ord, val,
       lead(val) OVER w AS lead,
       lead_ignore_nulls(val) OVER w AS lead_in,
       lead_ignore_nulls(val, 2) OVER w AS lead_in_off,
       lead_ignore_nulls(val, 2, 9999999) OVER w AS lead_in_off_d
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;
 part | ord |  val  | lead  | lead_in | lead_in_off | lead_in_off_d 
------+-----+-------+-------+---------+-------------+---------------
    1 |   1 | 64664 |  8779 |    8779 |       14005 |         14005
    1 |   2 |  8779 | 14005 |   14005 |       57699 |         57699
    1 |   3 | 14005 | 57699 |   57699 |       98842 |         98842
    1 |   4 | 57699 | 98842 |   98842 |       88563 |         88563
    1 |   5 | 98842 | 88563 |   88563 |       70453 |         70453
    1 |   6 | 88563 | 70453 |   70453 |       82824 |         82824
    1 |   7 | 70453 | 82824 |   82824 |       62453 |         62453
    1 |   8 | 82824 | 62453 |   62453 |             |       9999999
    1 |   9 | 62453 |       |         |             |       9999999
    2 |   1 |       | 51714 |   51714 |       17096 |         17096
    2 |   2 | 51714 | 17096 |   17096 |       41605 |         41605
    2 |   3 | 17096 | 41605 |   41605 |       15366 |         15366
    2 |   4 | 41605 | 15366 |   15366 |       87359 |         87359
    2 |   5 | 15366 | 87359 |   87359 |       98990 |         98990
    2 |   6 | 87359 | 98990 |   98990 |       34982 |         34982
    2 |   7 | 98990 | 34982 |   34982 |        3343 |          3343
    2 |   8 | 34982 |  3343 |    3343 |             |       9999999
    2 |   9 |  3343 |       |         |             |       9999999
    3 |   1 | 21903 | 24605 |   24605 |        6242 |          6242
    3 |   2 | 24605 |  6242 |    6242 |       24947 |         24947
    3 |   3 |  6242 | 24947 |   24947 |       79535 |         79535
    3 |   4 | 24947 | 79535 |   79535 |       66903 |         66903
    3 |   5 | 79535 | 66903 |   66903 |       42269 |         42269
    3 |   6 | 66903 | 42269 |   42269 |       31143 |         31143
    3 |   7 | 42269 | 31143 |   31143 |             |       9999999
    3 |   8 | 31143 |       |         |             |       9999999
    3 |   9 |       |       |         |             |       9999999
    4 |   1 |       | 49723 |   49723 |       23958 |         23958
    4 |   2 | 49723 | 23958 |   23958 |       80796 |         80796
    4 |   3 | 23958 | 80796 |   80796 |       41066 |         41066
    4 |   4 | 80796 |       |   41066 |       72991 |         72991
    4 |   5 |       | 41066 |   41066 |       72991 |         72991
    4 |   6 | 41066 | 72991 |   72991 |       33734 |         33734
    4 |   7 | 72991 | 33734 |   33734 |             |       9999999
    4 |   8 | 33734 |       |         |             |       9999999
    4 |   9 |       |       |         |             |       9999999
    5 |   1 |       |       |         |             |       9999999
    5 |   2 |       |       |         |             |       9999999
    5 |   3 |       |       |         |             |       9999999
    5 |   4 |       |       |         |             |       9999999
    5 |   5 |       |       |         |             |       9999999
    5 |   6 |       |       |         |             |       9999999
    5 |   7 |       |       |         |             |       9999999
    5 |   8 |       |       |         |             |       9999999
    5 |   9 |       |       |         |             |       9999999
(45 rows)

/* FIRST_VALUE */
SELECT part, ord, val,
       first_value(val) OVER w AS fv,
       first_value_ignore_nulls(val) OVER w AS fv_in,
       first_value_ignore_nulls(val, 9999999) OVER w AS fv_in_d
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;
 part | ord |  val  |  fv   | fv_in | fv_in_d 
------+-----+-------+-------+-------+---------
    1 |   1 | 64664 | 64664 | 64664 |   64664
    1 |   2 |  8779 | 64664 | 64664 |   64664
    1 |   3 | 14005 | 64664 | 64664 |   64664
    1 |   4 | 57699 |  8779 |  8779 |    8779
    1 |   5 | 98842 | 14005 | 14005 |   14005
    1 |   6 | 88563 | 57699 | 57699 |   57699
    1 |   7 | 70453 | 98842 | 98842 |   98842
    1 |   8 | 82824 | 88563 | 88563 |   88563
    1 |   9 | 62453 | 70453 | 70453 |   70453
    2 |   1 |       |       | 51714 |   51714
    2 |   2 | 51714 |       | 51714 |   51714
    2 |   3 | 17096 |       | 51714 |   51714
    2 |   4 | 41605 | 51714 | 51714 |   51714
    2 |   5 | 15366 | 17096 | 17096 |   17096
    2 |   6 | 87359 | 41605 | 41605 |   41605
    2 |   7 | 98990 | 15366 | 15366 |   15366
    2 |   8 | 34982 | 87359 | 87359 |   87359
    2 |   9 |  3343 | 98990 | 98990 |   98990
    3 |   1 | 21903 | 21903 | 21903 |   21903
    3 |   2 | 24605 | 21903 | 21903 |   21903
    3 |   3 |  6242 | 21903 | 21903 |   21903
    3 |   4 | 24947 | 24605 | 24605 |   24605
    3 |   5 | 79535 |  6242 |  6242 |    6242
    3 |   6 | 66903 | 24947 | 24947 |   24947
    3 |   7 | 42269 | 79535 | 79535 |   79535
    3 |   8 | 31143 | 66903 | 66903 |   66903
    3 |   9 |       | 42269 | 42269 |   42269
    4 |   1 |       |       | 49723 |   49723
    4 |   2 | 49723 |       | 49723 |   49723
    4 |   3 | 23958 |       | 49723 |   49723
    4 |   4 | 80796 | 49723 | 49723 |   49723
    4 |   5 |       | 23958 | 23958 |   23958
    4 |   6 | 41066 | 80796 | 80796 |   80796
    4 |   7 | 72991 |       | 41066 |   41066
    4 |   8 | 33734 | 41066 | 41066 |   41066
    4 |   9 |       | 72991 | 72991 |   72991
    5 |   1 |       |       |       | 9999999
    5 |   2 |       |       |       | 9999999
    5 |   3 |       |       |       | 9999999
    5 |   4 |       |       |       | 9999999
    5 |   5 |       |       |       | 9999999
    5 |   6 |       |       |       | 9999999
    5 |   7 |       |       |       | 9999999
    5 |   8 |       |       |       | 9999999
    5 |   9 |       |       |       | 9999999
(45 rows)

/* LAST_VALUE */
SELECT part, ord, val,
       last_value(val) OVER w AS lv,
       last_value_ignore_nulls(val) OVER w AS lv_in,
       last_value_ignore_nulls(val, -9999999) OVER w AS lv_in_d
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;
 part | ord |  val  |  lv   | lv_in | lv_in_d  
------+-----+-------+-------+-------+----------
    1 |   1 | 64664 | 14005 | 14005 |    14005
    1 |   2 |  8779 | 57699 | 57699 |    57699
    1 |   3 | 14005 | 98842 | 98842 |    98842
    1 |   4 | 57699 | 88563 | 88563 |    88563
    1 |   5 | 98842 | 70453 | 70453 |    70453
    1 |   6 | 88563 | 82824 | 82824 |    82824
    1 |   7 | 70453 | 62453 | 62453 |    62453
    1 |   8 | 82824 | 62453 | 62453 |    62453
    1 |   9 | 62453 | 62453 | 62453 |    62453
    2 |   1 |       | 17096 | 17096 |    17096
    2 |   2 | 51714 | 41605 | 41605 |    41605
    2 |   3 | 17096 | 15366 | 15366 |    15366
    2 |   4 | 41605 | 87359 | 87359 |    87359
    2 |   5 | 15366 | 98990 | 98990 |    98990
    2 |   6 | 87359 | 34982 | 34982 |    34982
    2 |   7 | 98990 |  3343 |  3343 |     3343
    2 |   8 | 34982 |  3343 |  3343 |     3343
    2 |   9 |  3343 |  3343 |  3343 |     3343
    3 |   1 | 21903 |  6242 |  6242 |     6242
    3 |   2 | 24605 | 24947 | 24947 |    24947
    3 |   3 |  6242 | 79535 | 79535 |    79535
    3 |   4 | 24947 | 66903 | 66903 |    66903
    3 |   5 | 79535 | 42269 | 42269 |    42269
    3 |   6 | 66903 | 31143 | 31143 |    31143
    3 |   7 | 42269 |       | 31143 |    31143
    3 |   8 | 31143 |       | 31143 |    31143
    3 |   9 |       |       | 31143 |    31143
    4 |   1 |       | 23958 | 23958 |    23958
    4 |   2 | 49723 | 80796 | 80796 |    80796
    4 |   3 | 23958 |       | 80796 |    80796
    4 |   4 | 80796 | 41066 | 41066 |    41066
    4 |   5 |       | 72991 | 72991 |    72991
    4 |   6 | 41066 | 33734 | 33734 |    33734
    4 |   7 | 72991 |       | 33734 |    33734
    4 |   8 | 33734 |       | 33734 |    33734
    4 |   9 |       |       | 33734 |    33734
    5 |   1 |       |       |       | -9999999
    5 |   2 |       |       |       | -9999999
    5 |   3 |       |       |       | -9999999
    5 |   4 |       |       |       | -9999999
    5 |   5 |       |       |       | -9999999
    5 |   6 |       |       |       | -9999999
    5 |   7 |       |       |       | -9999999
    5 |   8 |       |       |       | -9999999
    5 |   9 |       |       |       | -9999999
(45 rows)

/* NTH_VALUE */
SELECT part, ord, val,
       nth_value(val, 3) OVER w AS nth,
       nth_value_ignore_nulls(val, 3) OVER w AS nth_in
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;
 part | ord |  val  |  nth  | nth_in 
------+-----+-------+-------+--------
    1 |   1 | 64664 | 14005 |  14005
    1 |   2 |  8779 | 14005 |  14005
    1 |   3 | 14005 | 14005 |  14005
    1 |   4 | 57699 | 57699 |  57699
    1 |   5 | 98842 | 98842 |  98842
    1 |   6 | 88563 | 88563 |  88563
    1 |   7 | 70453 | 70453 |  70453
    1 |   8 | 82824 | 82824 |  82824
    1 |   9 | 62453 | 62453 |  62453
    2 |   1 |       | 17096 |       
    2 |   2 | 51714 | 17096 |  41605
    2 |   3 | 17096 | 17096 |  41605
    2 |   4 | 41605 | 41605 |  41605
    2 |   5 | 15366 | 15366 |  15366
    2 |   6 | 87359 | 87359 |  87359
    2 |   7 | 98990 | 98990 |  98990
    2 |   8 | 34982 | 34982 |  34982
    2 |   9 |  3343 |  3343 |   3343
    3 |   1 | 21903 |  6242 |   6242
    3 |   2 | 24605 |  6242 |   6242
    3 |   3 |  6242 |  6242 |   6242
    3 |   4 | 24947 | 24947 |  24947
    3 |   5 | 79535 | 79535 |  79535
    3 |   6 | 66903 | 66903 |  66903
    3 |   7 | 42269 | 42269 |  42269
    3 |   8 | 31143 | 31143 |  31143
    3 |   9 |       |       |       
    4 |   1 |       | 23958 |       
    4 |   2 | 49723 | 23958 |  80796
    4 |   3 | 23958 | 23958 |  80796
    4 |   4 | 80796 | 80796 |  80796
    4 |   5 |       |       |  41066
    4 |   6 | 41066 | 41066 |  72991
    4 |   7 | 72991 | 72991 |  33734
    4 |   8 | 33734 | 33734 |  33734
    4 |   9 |       |       |       
    5 |   1 |       |       |       
    5 |   2 |       |       |       
    5 |   3 |       |       |       
    5 |   4 |       |       |       
    5 |   5 |       |       |       
    5 |   6 |       |       |       
    5 |   7 |       |       |       
    5 |   8 |       |       |       
    5 |   9 |       |       |       
(45 rows)

SELECT part, ord, val,
       nth_value(val, 3) OVER w AS nth,
       nth_value_from_last(val, 3) OVER w AS nth_fl
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;
 part | ord |  val  |  nth  | nth_fl 
------+-----+-------+-------+--------
    1 |   1 | 64664 | 14005 |  64664
    1 |   2 |  8779 | 14005 |   8779
    1 |   3 | 14005 | 14005 |  14005
    1 |   4 | 57699 | 57699 |  57699
    1 |   5 | 98842 | 98842 |  98842
    1 |   6 | 88563 | 88563 |  88563
    1 |   7 | 70453 | 70453 |  70453
    1 |   8 | 82824 | 82824 |  70453
    1 |   9 | 62453 | 62453 |  70453
    2 |   1 |       | 17096 |       
    2 |   2 | 51714 | 17096 |  51714
    2 |   3 | 17096 | 17096 |  17096
    2 |   4 | 41605 | 41605 |  41605
    2 |   5 | 15366 | 15366 |  15366
    2 |   6 | 87359 | 87359 |  87359
    2 |   7 | 98990 | 98990 |  98990
    2 |   8 | 34982 | 34982 |  98990
    2 |   9 |  3343 |  3343 |  98990
    3 |   1 | 21903 |  6242 |  21903
    3 |   2 | 24605 |  6242 |  24605
    3 |   3 |  6242 |  6242 |   6242
    3 |   4 | 24947 | 24947 |  24947
    3 |   5 | 79535 | 79535 |  79535
    3 |   6 | 66903 | 66903 |  66903
    3 |   7 | 42269 | 42269 |  42269
    3 |   8 | 31143 | 31143 |  42269
    3 |   9 |       |       |  42269
    4 |   1 |       | 23958 |       
    4 |   2 | 49723 | 23958 |  49723
    4 |   3 | 23958 | 23958 |  23958
    4 |   4 | 80796 | 80796 |  80796
    4 |   5 |       |       |       
    4 |   6 | 41066 | 41066 |  41066
    4 |   7 | 72991 | 72991 |  72991
    4 |   8 | 33734 | 33734 |  72991
    4 |   9 |       |       |  72991
    5 |   1 |       |       |       
    5 |   2 |       |       |       
    5 |   3 |       |       |       
    5 |   4 |       |       |       
    5 |   5 |       |       |       
    5 |   6 |       |       |       
    5 |   7 |       |       |       
    5 |   8 |       |       |       
    5 |   9 |       |       |       
(45 rows)

SELECT part, ord, val,
       nth_value_from_last(val, 3) OVER w AS nth_fl,
       nth_value_from_last_ignore_nulls(val, 3) OVER w AS nth_fl_in
FROM things
WINDOW w AS (PARTITION BY part ORDER BY ord ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING)
ORDER BY part, ord;
 part | ord |  val  | nth_fl | nth_fl_in 
------+-----+-------+--------+-----------
    1 |   1 | 64664 |  64664 |     64664
    1 |   2 |  8779 |   8779 |      8779
    1 |   3 | 14005 |  14005 |     14005
    1 |   4 | 57699 |  57699 |     57699
    1 |   5 | 98842 |  98842 |     98842
    1 |   6 | 88563 |  88563 |     88563
    1 |   7 | 70453 |  70453 |     70453
    1 |   8 | 82824 |  70453 |     70453
    1 |   9 | 62453 |  70453 |     70453
    2 |   1 |       |        |          
    2 |   2 | 51714 |  51714 |     51714
    2 |   3 | 17096 |  17096 |     17096
    2 |   4 | 41605 |  41605 |     41605
    2 |   5 | 15366 |  15366 |     15366
    2 |   6 | 87359 |  87359 |     87359
    2 |   7 | 98990 |  98990 |     98990
    2 |   8 | 34982 |  98990 |     98990
    2 |   9 |  3343 |  98990 |     98990
    3 |   1 | 21903 |  21903 |     21903
    3 |   2 | 24605 |  24605 |     24605
    3 |   3 |  6242 |   6242 |      6242
    3 |   4 | 24947 |  24947 |     24947
    3 |   5 | 79535 |  79535 |     79535
    3 |   6 | 66903 |  66903 |     66903
    3 |   7 | 42269 |  42269 |     66903
    3 |   8 | 31143 |  42269 |     66903
    3 |   9 |       |  42269 |          
    4 |   1 |       |        |          
    4 |   2 | 49723 |  49723 |     49723
    4 |   3 | 23958 |  23958 |     49723
    4 |   4 | 80796 |  80796 |     23958
    4 |   5 |       |        |     80796
    4 |   6 | 41066 |  41066 |     41066
    4 |   7 | 72991 |  72991 |     41066
    4 |   8 | 33734 |  72991 |     41066
    4 |   9 |       |  72991 |          
    5 |   1 |       |        |          
    5 |   2 |       |        |          
    5 |   3 |       |        |          
    5 |   4 |       |        |          
    5 |   5 |       |        |          
    5 |   6 |       |        |          
    5 |   7 |       |        |          
    5 |   8 |       |        |          
    5 |   9 |       |        |          
(45 rows)

DROP TABLE things;
DROP EXTENSION extra_window_functions;
